Dockerfile

```diff
@@ -0,0 +1,44 @@
+FROM ubuntu:bionic as source
+
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y \
+    git ca-certificates && \
+    rm -rf /var/lib/apt/lists/*
+
+COPY . selfoss/
+
+FROM composer:1.9 as composer
+
+COPY --from=source /selfoss /selfoss
+
+RUN cd /selfoss && composer install --ignore-platform-reqs --optimize-autoloader --no-dev
+
+
+
+FROM node:13-buster as npm
+
+COPY --from=composer /selfoss /selfoss
+
+RUN cd /selfoss && npm i && npm install --prefix assets/ && npm run build
```

Comment:
```suggestion
RUN cd /selfoss && npm install && npm install --prefix assets/ && npm run build
```

Though the top-level `npm install` is only used for creating a dist tarball so maybe we can omit it altogether.

```suggestion
RUN cd /selfoss && npm install --prefix assets/ && npm run build
```

Reply:
> Though the top-level `npm install` is only used for creating a dist tarball so maybe we can omit it altogether.

Done the updates

Reply:
> Though the top-level `npm install` is only used for creating a dist tarball so maybe we can omit it altogether.

So, we can remove the npm install part altogether? please correct me if I'm wrong because I didn't get it properly

Reply:
Well, the top-level `npm install` only installs `archiver` needed for `npm run dist`:

https://github.com/SSilence/selfoss/blob/cdde3b5affb931d360f4952b3f2b6d15cdaf467a/package.json#L9-L11

But `npm` also runs `postinstall` script:

https://github.com/SSilence/selfoss/blob/cdde3b5affb931d360f4952b3f2b6d15cdaf467a/package.json#L32

which will install the dependencies:

https://github.com/SSilence/selfoss/blob/cdde3b5affb931d360f4952b3f2b6d15cdaf467a/package.json#L27-L29

So the first `install` will install `archiver` that we do not need but also run the second `install` so we are running it twice. And it will also run `composer install`, I wonder how it does not fail.

Reply:
> Well, the top-level `npm install` only installs `archiver` needed for `npm run dist`:
> 
> https://github.com/SSilence/selfoss/blob/cdde3b5affb931d360f4952b3f2b6d15cdaf467a/package.json#L9-L11
> 
> But `npm` also runs `postinstall` script:
> 
> https://github.com/SSilence/selfoss/blob/cdde3b5affb931d360f4952b3f2b6d15cdaf467a/package.json#L32
> 
> which will install the dependencies:
> 
> https://github.com/SSilence/selfoss/blob/cdde3b5affb931d360f4952b3f2b6d15cdaf467a/package.json#L27-L29
> 
> So the first `install` will install `archiver` that we do not need but also run the second `install` so we are running it twice. And it will also run `composer install`, I wonder how it does not fail.

Not sure. we are not running run npm install-dependencies because it is failing.
Not only on docker, even on my LAMP server which is based on Ubuntu 18.04 and on XAMPP also.
I've checked it from master branch too, but there also the same issue.
in october it aas working fine.

Reply:
> Not sure. we are not running run npm install-dependencies because it is failing.
> Not only on docker, even on my LAMP server which is based on Ubuntu 18.04 and on XAMPP also.
> I've checked it from master branch too, but there also the same issue.
> in october it aas working fine.

Could you share the error message? It works for me here on master.

Reply:
> > Not sure. we are not running run npm install-dependencies because it is failing.
> > Not only on docker, even on my LAMP server which is based on Ubuntu 18.04 and on XAMPP also.
> > I've checked it from master branch too, but there also the same issue.
> > in october it aas working fine.
> 
> Could you share the error message? It works for me here on master.

Sure, here it is:
(I don't know whether the screenshot is showing properly or not as I'm a visually impaired user). If it doesn't display properly, then please let me know
![image](https://user-images.githubusercontent.com/6381747/71484536-ea252480-2832-11ea-8775-0adcf6b56e8b.png)

Reply:
This was from calling `npm run install-dependencies`? The only thing I see is the fist line: Missing package version could you try adding `version` attribute to assets/package.json. Though I wonder why does not trip it my npm of the almost same version (6.13.0).

Reply:
> This was from calling `npm run install-dependencies`? The only thing I see is the fist line: Missing package version could you try adding `version` attribute to assets/package.json. Though I wonder why does not trip it my npm of the almost same version (6.13.0).

Yes, I ran npm run install-dependencies. so, should I add npm package version to assets/package.json?

Reply:
I would try

```patch
--- a/assets/package.json
+++ b/assets/package.json
@@ -1,4 +1,5 @@
 {
+  "name": "selfoss-assets",
   "dependencies": {
     "@fancyapps/fancybox": "^3.1.20",
     "@fortawesome/fontawesome-svg-core": "^1.2.25",

```

but it is weird. According to [this code](https://github.com/npm/cli/blob/44ddd0b528732028ea4f38ef15b5826f3f4e2ec1/lib/install/validate-args.js#L24-L32), it should ask for name first. Unless it comes from the top-level `package.json`.

What do you get by running `npm run install-dependencies > output.txt 2>&1`? I get [output.txt](https://github.com/SSilence/selfoss/files/4004269/output.txt).

Reply:
> I still get the same error, tried reinstalling node but no luck. :(
> don't know what's the root of this issue.
> Github is not allowing me to upload the output file so below's the pastebin URL:
> https://pastebin.com/3cJ46piE



> I would try
> 
> ```diff
> --- a/assets/package.json
> +++ b/assets/package.json
> @@ -1,4 +1,5 @@
>  {
> +  "name": "selfoss-assets",
>    "dependencies": {
>      "@fancyapps/fancybox": "^3.1.20",
>      "@fortawesome/fontawesome-svg-core": "^1.2.25",
> ```
> 
> but it is weird. According to [this code](https://github.com/npm/cli/blob/44ddd0b528732028ea4f38ef15b5826f3f4e2ec1/lib/install/validate-args.js#L24-L32), it should ask for name first. Unless it comes from the top-level `package.json`.
> 
> What do you get by running `npm run install-dependencies > output.txt 2>&1`? I get [output.txt](https://github.com/SSilence/selfoss/files/4004269/output.txt).

Reply:
Weird. At this point I would pin it as some npm on Windows idiosyncrasy. Does running the command `npm install --production=false --prefix assets/` directly also fail? What about `cd assets && npm install --production=false`?

Reply:
> I'll check with these commands too.
> but it is not related to Windows because same error comes while in the docker container too and on my linux server as well.
I think that it is related to node version? which version you are using?


> Weird. At this point I would pin it as some npm on Windows idiosyncrasy. Does running the command `npm install --production=false --prefix assets/` directly also fail? What about `cd assets && npm install --production=false`?

Reply:
I tried node 10.17.0 with npm 6.13.0, node 10.18.0 with npm 6.13.4, and node 13.3.0 with 6.13.1.

Reply:
I think that I should try it with node version 13.3.0 then, may be it will work. 

> I tried node 10.17.0 with npm 6.13.0, node 10.18.0 with npm 6.13.4, and node 13.3.0 with 6.13.1.